### ISSUE TITLE
End handler does not seem to be called if the json does not end

### ISSUE THREAD

**Issue by TheBlue-1** [xtao-org/jsonhilo#6]
When the calling end from the outside without parsing a whole json the event is not passed to the handler.

Example:
```ts
 type RequiredHandlers = Required<JsonHighHandlers<unknown, unknown>>;
    const handlers: RequiredHandlers = {
      openArray: () => {
        console.log("openArray");
      },
      openObject: () => {
        console.log("openObject");
      },
      closeArray: () => {
        console.log("closeArray");
      },
      closeObject: () => {
        console.log("closeObject");
      },
      key: (key) => {
        console.log("key", key);
      },
      value: (value) => {
        console.log("value", value);
      },
      end: () => {
        console.log("end");
      },
    };

    const streamLike = JsonHigh(handlers);

    const writable = new WritableStream({
      write: (chunk) => {
        try {
          const decoded = new TextDecoder().decode(chunk as Uint8Array);
          streamLike.chunk(decoded);
        } catch (e) {
          throw new Error(`Error while streaming.`, { cause: e });
        }
      },
      close: () => {
        console.log("close");
        streamLike.end();
      },
      abort: () => {
        console.log("abort");
        console.warn("unexpected abort received");
        streamLike.end();
      },
    });
    const abortController = new AbortController();

    const readable = new ReadableStream({
      start(controller) {
        controller.enqueue(new TextEncoder().encode('{"a":2,"b":3')); // works with '{"a":2,"b":3}'
        controller.close();
      },
    });

    await readable.pipeTo(writable, { signal: abortController.signal });
    console.log("done");
```
This example will have the output 
```
openObject
key a
value 2
key b
close
done
```
where "close" shows that `end` got called
when parsing a whole json like  `{"a":2,"b":3}` the end function is called and the output looks like
```
openObject
key a
value 2
key b
value 3
closeObject
close
end
done
```

I would expect that the end event is also allowed to be used without reading a whole valid json.

**Comment by djedr** [comment]
Thank you for reporting the issue! :slightly_smiling_face: 

There is indeed a problem here, but the fact that the end handler is not called is by design (although the design might benefit from improving and a bit more documentation :wink:).

I think the simplest improvement which will minimally affect backwards compatibility would be to throw an exception from `JsonHigh` when `streamLike.end()` is called on incomplete JSON. Gotta think a bit more to be sure. Let me know if that would work for you.


Currently a handler is called only if there was no error upstream of it, otherwise it isn't and the error is propagated through return values.

<details>

 `JsonLow` (the underlying low-level parser) calls the downstream end handler only if there is no error, if the JSON is complete (i.e. there are no unclosed objects/arrays/strings) -- in that case  whatever the handler returned will be passed upstream. Otherwise, if there is an error (i.e. JSON is incomplete) `JsonLow` returns an error object instead, without notifying downstream. This is consistent with what happens when there is an error somewhere before the end.

</details>

You can see the error object if you `console.log` the return value of `streamLike.end()`:

```js
  close: () => {
    console.log("close");
    console.log(streamLike.end());
  },
```

With the proposed simple improvement you would get an exception instead. This would make `JsonHigh.end()` consistent with `JsonHigh.chunk()` which also throws exceptions if there is an error downstream.

An explanatory note: one important reason for avoiding exceptions in favor of this error propagation mechanism was performance.

**Comment by TheBlue-1** [comment]
Thank you for the quick answer. 
I actually think that throwing an error could cause more problems than it fixes 🤐
So probably it would be enough to adjust the docs.

Whats the reason that the end-handler is not called if the json is not complete? I would think the handler can be used for cleanup-like things which should also be done if a json is not complete.

**Comment by djedr** [comment]
> I actually think that throwing an error could cause more problems than it fixes 🤐

Yes, it would break backward-compatibility, so it would be version 0.4.0.

> So probably it would be enough to adjust the docs.

Indeed, I think I'll do that for now.

> Whats the reason that the end-handler is not called if the json is not complete?

Same reason any other handler is not called if the parser receives an unexpected code point.

Calling a handler implies that the parser has recognized a code point. The end signal can be seen as a special code point that signifies the end of the JSON document being analyzed. Calling `.end()` is basically like calling `.codePoint(EOF)`.

<details>

The logic goes like this.

If the JSON grammar allows a particular code point at the current position, it is recognized and an appropriate downstream handler is called. Whatever that handler returns is returned upstream.

Otherwise, when a code point is not recognized, there is no handler to call. Instead the parser returns an error upstream.

When you call `streamLike.end()`, you are telling the parser "the current JSON document ends here". If the JSON grammar does not allow the document to end at that point, it's analogous to an invalid code point. So the parser returns an error.

Calling the end handler would signal that the end was recognized/allowed, which would be incorrect.

</details>

> I would think the handler can be used for cleanup-like things which should also be done if a json is not complete.

So basically you need something like a `finally` clause. :wink: 

I think something like this would work:

```
const ret = streamLike.end()
if (isError(ret)) handle(ret)
cleanup()
```

Unless `isError` or `handle` could throw. In that case it would be:

```
const ret = stream.end()
try { if (isError(ret)) handle(ret) } 
// optional catch (e) { ... } here
finally { cleanup() }
```

This should cover your use case.

`isError` would be something like:

```
const isError = (value) => {
  // JsonFeedbackType can be imported from '@xtao-org/jsonhilo'
  return value !== null && typeof value === 'object' && value.type === JsonFeedbackType.error
}
```

This check is somewhat awkward (this is where the design could use some improving, I'll think about that), but that should do it.

An alternative (and a more conventional) way of signaling errors, would be an `.error()` handler. That may be cleaner for your use case. I'll think about it for `0.4.0`.

Hope this helps!

Cheerio :slightly_smiling_face:

**Comment by TheBlue-1** [comment]
Thank you

**Comment by djedr** [comment]
Merged #7, released v0.3.7 (https://www.npmjs.com/package/@xtao-org/jsonhilo/v/0.3.7). Added docs to README and the `isError` helper from the above comment for convenience. Have a good one!
